src/common/HordeAd/Internal/TensorOps.hs

```diff
@@ -128,31 +136,29 @@ tindexR
   => OR.Array (1 + n) r -> Int -> OR.Array n r
 tindexR = OR.index
 
+-- TODO: optimize to tindexR for n == 0
 tindex0R
   :: Numeric r
-  => OR.Array n r -> [Int] -> r
+  => OR.Array n r -> IndexInt n -> r
 tindex0R (Data.Array.Internal.RankedS.A
             (Data.Array.Internal.RankedG.A _
                Data.Array.Internal.T{..})) is =
-  values V.! (offset + sum (zipWith (*) is strides))
+  values V.! (offset + sum (zipWith (*) (indexToList is) strides))
+    -- to avoid linearizing @values@, we do everything in unsized way
 
 -- TODO: optimize to tindex0R for n == 0
 tindexNR
-  :: (KnownNat n, Numeric r)
-  => OR.Array (1 + m + n) r -> [Int] -> OR.Array n r
-tindexNR arr ixs =
-  let Data.Array.Internal.DynamicS.A
-        (Data.Array.Internal.DynamicG.A sh
-           Data.Array.Internal.T{..}) =
-             OT.normalize $ Data.Array.Convert.convert arr
-               -- OT to avoid KnownNat m, which breaks typing of other code
-               -- due to no sized lists
+  :: forall m n r. (KnownNat m, KnownNat n, Numeric r)
+  => OR.Array (m + n) r -> IndexInt m -> OR.Array n r
+tindexNR arr ix =
+  let Data.Array.Internal.RankedS.A
+        (Data.Array.Internal.RankedG.A sh
+           Data.Array.Internal.T{offset, values}) = OR.normalize arr
       !_A = assert (offset == 0) ()
-  in let pathToIx is = sum (zipWith (*) is strides)
-         ix = pathToIx ixs
-         shN = drop (length ixs) sh
+  in let i = toLinearIdx @m @n (listShapeToShape sh) ix
+         shN = drop (valueOf @m) sh
          len = product shN
-     in OR.fromVector shN $ V.slice ix len values
+     in OR.fromVector shN $ V.slice i len values
```

Comment:
You are probably right on all accounts. Let's test...

Reply:
No, it seems you were mistaken about it being wrong. Tests pass. Swapped. I wonder if 

https://github.com/Mikolaj/horde-ad/blob/8e49146b893ad1ed238722326924c9688079ce5f/src/common/HordeAd/Internal/TensorOps.hs#L88-L102

can be simplified and sped up similarly.

Reply:
BTW, the new implementation of `tindexNR` if funny in that when projecting a 1M-element tensor to a single Double (presented as a rank 0 tensor), we keep the whole underlying 1M `values` vector. But this is fine --- the vector is going to be shared between all such Doubles and, probably, may even be a perfect packed representation for a set of all of them, as opposed to allocating each, separately, on the heap (even as a Double, not a rank 0 tensor).

That differs from `updateNR, where we actually update the `values` vector, which rewrites is completely, so not even a portion of the original vector is shared.

Reply:
It works?! Yay for strided representations, I guess.

Regarding updateNR: will those vjoin operations be done in-place somehow? If not, I think re-allocating the entire tensor after every update will be a much bigger inefficiency than the flattining, which is just a _single_ re-allocation, after all.

I wonder if you could allocate an `m`-dimensional array of `Int`s, where `-1` indicates "keep as-is" and anything `>=0` indicates the index in the `upd` list of the tensor to put there. Then `build` the destination array in one go. The boolean array seems redundant allocation but it's much less than the current _repeated_ reallocation of the entire thing.

Reply:
With the new non-allocating `tindexNR` it might be useful to export something like [ByteString.copy](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString.html#v:copy) (semantically `id`) to destroy the sharing you mention. Sometimes it may be counterproductive.

Reply:
> Regarding updateNR: will those vjoin operations be done in-place somehow?

Right, they should be and probably will be in the future. But for now, I'm forgetting this is a list of updates and thinking about the simplest case of one-element list (which is the only one I'm using in horde-ad so far IIRC, incidentally). But you are right, context matters. Perhaps I should define a separate operation with just one update and it would be very similar to `tindexNR` and then try to optimize that separately.

Yes, using build for that makes some sense, especially for larger lists of larger tensors.

Reply:
Alternative to what I wrote above: or sort the `upd` list and construct one big list to be passed to `vjoin` _once_. That's actually probably the fastest one if the input arrays are sufficiently large and `upd` is not too long -- a bit of list manipulation is negligible compared to conditional code run elementwise over a huge array.

Reply:
Thanks. This may work. It's so fortunate subtensors can't overlap non-trivially. I will remember to look into these notes once I profile and see which  `updateNR` actually matters: the singleton one, or the general one.